### PR TITLE
Support for platforms based on Barefoot Networks' device

### PR DESCRIPTION
**Pull request by mkbalani**
**- Incorporate support for platforms based on Barefoot Networks' device **

**- Added port config files and related scripts for the platforms **

**- Verified using SONiC tests (requires the corresponding changes to sairedis and swss) **

p4 barefoot

### CODE REVIEW

**Comment by lguohan** [comment]
please remove all irrelevant related to your platform update.

**Comment by lguohan** [comment]
too many unrelated changes and conflicts.

**Comment by mkbalani** [comment]
Removed all the irrelevant changes from PR. Seems like squashing and merging was causing a issue in the diffs rendered. The PR now has only Barefoot specific changes.
Will update .gitmodules after swss sairedis and swss-common changes are merged.

File: .gitmodules [hunk 1]
```diff
@@ -1,15 +1,15 @@
 [submodule "sonic-swss-common"]
 	path = src/sonic-swss-common
-	url = https://github.com/Azure/sonic-swss-common
+	url = https://github.com/barefootnetworks/sonic-swss-common
```
Comment (lguohan) [inline .gitmodules]: cannot use barefootnetwork repo for sonic-swss-common

File: .gitmodules [hunk 2]
```diff
@@ -1,15 +1,15 @@
...
 [submodule "sonic-linux-kernel"]
 	path = src/sonic-linux-kernel
 	url = https://github.com/Azure/sonic-linux-kernel
 [submodule "sonic-sairedis"]
 	path = src/sonic-sairedis
-	url = https://github.com/Azure/sonic-sairedis
+	url = https://github.com/barefootnetworks/sonic-sairedis
```
Comment (lguohan) [inline .gitmodules]: cannot use barefootnetwork repo for sonic-swss-common

File: .gitmodules [hunk 3]
```diff
@@ -1,15 +1,15 @@
...
 [submodule "sonic-swss"]
 	path = src/sonic-swss
-	url = https://github.com/Azure/sonic-swss
+	url = https://github.com/barefootnetworks/sonic-swss
```
Comment (lguohan) [inline .gitmodules]: cannot use barefootnetwork repo for sonic-swss-common

File: files/image_config/interfaces/interfaces.j2 [hunk 4]
```diff
@@ -18,6 +18,10 @@ iface lo {{ 'inet' if prefix | ipv4 else 'inet6' }} static
 {% endfor %}
 {% endblock loopback %}
 {% block mgmt_interface %}
+# BMC interface
+auto usb0
+iface usb0 inet6 auto
```
Comment (lguohan) [inline files/image_config/interfaces/interfaces.j2]: what if there is no usb0?

File: platform/barefoot/bfn-platform-ingrasys.mk [hunk 5]
```diff
@@ -0,0 +1,5 @@
+BFN_INGRASYS_PLATFORM = bfnplatform-ingrasys_1.0.0_amd64.deb
+$(BFN_INGRASYS_PLATFORM)_URL = "https://github.com/Ingrasys-sonic/packages/raw/master/lib/bfnplatform-ingrasys_1.0.0_amd64.deb"
```
Comment (lguohan) [inline platform/barefoot/bfn-platform-ingrasys.mk]: why binary?

File: platform/barefoot/bfn-platform-wnc.mk [hunk 6]
```diff
@@ -0,0 +1,5 @@
+WNC_OSW1800_PLATFORM = bfnplatformwnc_1.0.0_amd64.deb
+$(WNC_OSW1800_PLATFORM)_URL = "https://github.com/YaoTien/download/raw/master/sonic/sde/7_0_0_18/bfnplatformwnc_1.0.0_amd64.deb"
```
Comment (lguohan) [inline platform/barefoot/bfn-platform-wnc.mk]: platform driver should be source code, why binary?

File: platform/barefoot/bfn-platform.mk [hunk 7]
```diff
@@ -0,0 +1,6 @@
+BFN_PLATFORM = bfnplatform_1.0.0_amd64.deb
+$(BFN_PLATFORM)_URL =	"https://github.com/barefootnetworks/sonic-release-pkgs/raw/rel_8_2/bfnplatform_1.0.0_amd64.deb"
```
Comment (lguohan) [inline platform/barefoot/bfn-platform.mk]: platform driver should be source code, why binary?

File: platform/barefoot/sonic-platform-modules-bfn/debian/control [hunk 8]
```diff
@@ -0,0 +1,12 @@
+Source: platform-modules-bfn
+Section: main
+Priority: extra
+Maintainer: support <support@edge-core.com>
+Build-Depends: debhelper (>= 8.0.0), bzip2
+Standards-Version: 3.9.3
+
+Package: platform-modules-bfn
+Architecture: amd64
+Depends: linux-image-3.16.0-4-amd64
```
Comment (lguohan) [inline platform/barefoot/sonic-platform-modules-bfn/debian/control]: 3.16.0-5-amd64

File: platform/barefoot/sonic-platform-modules-wnc-osw1800/debian/control [hunk 9]
```diff
@@ -0,0 +1,12 @@
+Source: platform-modules-wnc-osw1800
+Section: main
+Priority: extra
+Maintainer: WNC <wnc@wnc.com.tw>
+Build-Depends: debhelper (>= 8.0.0), bzip2
+Standards-Version: 3.9.3
+
+Package: platform-modules-wnc-osw1800
+Architecture: amd64
+Depends: linux-image-3.16.0-4-amd64
```
Comment (lguohan) [inline platform/barefoot/sonic-platform-modules-wnc-osw1800/debian/control]: 3.16.0-5-amd64

File: platform/barefoot/sonic-platform-modules-wnc-osw1800/scripts/device_node.sh [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+# MUX0 channel3
+CHANNEL=3
+echo wnc_eeprom 0x50 > /sys/bus/i2c/devices/i2c-$(($START_NODE_NUM+$CHANNEL))/new_device
+echo wnc_eeprom 0x51 > /sys/bus/i2c/devices/i2c-$(($START_NODE_NUM+$CHANNEL))/new_device
+
+# MUX0 channel4
+CHANNEL=4
+echo wnc_eeprom 0x54 > /sys/bus/i2c/devices/i2c-$(($START_NODE_NUM+$CHANNEL))/new_device
+echo tmp421 0x1E > /sys/bus/i2c/devices/i2c-$(($START_NODE_NUM+$CHANNEL))/new_device
+sleep 1
```
Comment (lguohan) [inline platform/barefoot/sonic-platform-modules-wnc-osw1800/scripts/device_node.sh]: why sleep here?

File: .gitmodules (hunk 1 above)
Comment (mkbalani) [reply .gitmodules]: This is just temporary until https://github.com/Azure/sonic-swss-common/pull/205 is merged. Will revert to azure repos before merging

File: .gitmodules (hunk 2 above)
Comment (mkbalani) [reply .gitmodules]: same as above

File: .gitmodules (hunk 3 above)
Comment (mkbalani) [reply .gitmodules]: same as above

Comment (lguohan) [reply .gitmodules]: merged.

File: .gitmodules (hunk 1 above)
Comment (lguohan) [reply .gitmodules]: merged

File: dockers/docker-base/Dockerfile.j2 [hunk 11]
```diff
@@ -29,6 +29,7 @@ RUN apt-get -y install \
     vim-tiny           \
     perl               \
     python             \
+    psmisc             \
```
Comment (lguohan) [inline dockers/docker-base/Dockerfile.j2]: why do we need this psmisc in the docker-base?

**Comment by lguohan** [comment]
please help to resolve conflict

File: dockers/docker-base/Dockerfile.j2 (hunk 11 above)
Comment (lguohan) [reply dockers/docker-base/Dockerfile.j2]: @pavel-shirshov , I believe sonic-utilties has psmisc dependency. 

anyway, @mkbalani  please remove this from this PR. If it is an issue, please raise separate PR for this.

File: files/image_config/interfaces/interfaces.j2 (hunk 4 above)
Comment (mkbalani) [reply files/image_config/interfaces/interfaces.j2]: If I understand correctly, I could remove this from here in the template file. But the output generated using cfggen with this file is going to be compared with a sample interfaces file which is fixed. If I keep the usb0 instance there and not here (conditionally), then sonic build shall itself fail.

File: dockers/docker-base/Dockerfile.j2 (hunk 11 above)
Comment (mkbalani) [reply dockers/docker-base/Dockerfile.j2]: removed for now

File: .gitmodules [hunk 12]
```diff
@@ -77,6 +77,9 @@
 [submodule "platform/broadcom/sonic-platform-modules-mitac"]
 	path = platform/broadcom/sonic-platform-modules-mitac
 	url = https://github.com/MiTAC-EBU/sonic-platform-modules-mitac.git
+[submodule "platform/barefoot/sonic-platform-modules-ingrasys"]
+	path = platform/barefoot/sonic-platform-modules-ingrasys
+	url = https://github.com/Ingrasys-sonic/sonic-platform-modules-ingrasys-barefoot.git
```
[the remark this answers is not in the document]
Comment (mkbalani) [reply .gitmodules]: done

**Comment by mkbalani** [comment]
@lguohan It seems the vs build job is still somehow using an old refpoint for swss. From the PR diff here [swss-ref](https://github.com/Azure/sonic-swss/compare/762e7da80142358af0d348dea1143f38aa60a992...dcf6c905410f08b004e880dbed56823d29e7bd5e)
I can see that the refpoint is correctly dcf6c, however the jenkins build shows below error `>       assert len(keys) == 1
E       AssertionError: assert 3 == 1
E        +  where 3 = len(('oid:0x700000000058a', 'oid:0x700000000058f', 'oid:0x700000000058e'))` which should not be seen with this refpoint

**Comment by lguohan** [comment]
retest this please

File: platform/barefoot/sonic-platform-modules-ingrasys/debian/sonic-platform-ingrasys-s9180-32x.postinst [hunk 13]
```diff
@@ -0,0 +1,57 @@
+# Automatically added by dh_systemd_enable
```
Comment (lguohan) [inline platform/barefoot/sonic-platform-modules-ingrasys/debian/sonic-platform-ingrasys-s9180-32x.postinst]: i think this is automatically added by debian, remove from the check-in.

File: platform/barefoot/sonic-platform-modules-ingrasys/debian/sonic-platform-ingrasys-s9180-32x.install [hunk 14]
```diff
@@ -0,0 +1,4 @@
+lib/systemd/
```
Comment (lguohan) [inline platform/barefoot/sonic-platform-modules-ingrasys/debian/sonic-platform-ingrasys-s9180-32x.install]: i think this is automatically added by debian, remove from the check-in.

File: platform/barefoot/sonic-platform-modules-ingrasys/debian/sonic-platform-ingrasys-s9180-32x.dirs [hunk 15]
```diff
@@ -0,0 +1,4 @@
+usr/sbin
```
Comment (lguohan) [inline platform/barefoot/sonic-platform-modules-ingrasys/debian/sonic-platform-ingrasys-s9180-32x.dirs]: i think this is automatically added by debian, remove from the check-in.

File: platform/barefoot/sonic-platform-modules-ingrasys/debian/sonic-platform-ingrasys-s9180-32x.postrm [hunk 16]
```diff
@@ -0,0 +1,42 @@
+# Automatically added by dh_systemd_start
```
Comment (lguohan) [inline platform/barefoot/sonic-platform-modules-ingrasys/debian/sonic-platform-ingrasys-s9180-32x.postrm]: i think this is automatically added by debian, remove from the check-in.

File: platform/barefoot/sonic-platform-modules-ingrasys/debian/sonic-platform-ingrasys-s9180-32x.prerm [hunk 17]
```diff
@@ -0,0 +1,21 @@
+# Automatically added by dh_systemd_start
```
Comment (lguohan) [inline platform/barefoot/sonic-platform-modules-ingrasys/debian/sonic-platform-ingrasys-s9180-32x.prerm]: this file is automatically added by debian, remove from the check-in.

File: files/image_config/interfaces/interfaces.j2 (hunk 4 above)
Comment (lguohan) [reply files/image_config/interfaces/interfaces.j2]: then you can remove the usb0 in the cfggen sample interface file.

**Comment by lguohan** [comment]
retest this please

**Comment by mkbalani** [comment]
retesting build and fixes

**Comment by mkbalani** [comment]
@lguohan BTW azure swss master is still pointing to 762e7da, rather than dcf6c, because of which vs job is failing I think.

File: files/image_config/interfaces/interfaces.j2 (hunk 4 above)
Comment (mkbalani) [reply files/image_config/interfaces/interfaces.j2]: fixed

File: platform/barefoot/sonic-platform-modules-ingrasys/debian/sonic-platform-ingrasys-s9180-32x.postinst (hunk 13 above)
Comment (mkbalani) [reply platform/barefoot/sonic-platform-modules-ingrasys/debian/sonic-platform-ingrasys-s9180-32x.postinst]: removed file

File: platform/barefoot/sonic-platform-modules-ingrasys/debian/sonic-platform-ingrasys-s9180-32x.install (hunk 14 above)
Comment (mkbalani) [reply platform/barefoot/sonic-platform-modules-ingrasys/debian/sonic-platform-ingrasys-s9180-32x.install]: removed file

File: platform/barefoot/sonic-platform-modules-ingrasys/debian/sonic-platform-ingrasys-s9180-32x.dirs (hunk 15 above)
Comment (mkbalani) [reply platform/barefoot/sonic-platform-modules-ingrasys/debian/sonic-platform-ingrasys-s9180-32x.dirs]: removed file

File: platform/barefoot/sonic-platform-modules-ingrasys/debian/sonic-platform-ingrasys-s9180-32x.postrm (hunk 16 above)
Comment (mkbalani) [reply platform/barefoot/sonic-platform-modules-ingrasys/debian/sonic-platform-ingrasys-s9180-32x.postrm]: removed file

File: platform/barefoot/sonic-platform-modules-ingrasys/debian/sonic-platform-ingrasys-s9180-32x.prerm (hunk 17 above)
Comment (mkbalani) [reply platform/barefoot/sonic-platform-modules-ingrasys/debian/sonic-platform-ingrasys-s9180-32x.prerm]: removed file

Comment (mkbalani) [reply platform/barefoot/sonic-platform-modules-ingrasys/debian/sonic-platform-ingrasys-s9180-32x.prerm]: BFN build break after removing this file. Adding it back for now. We shall have ingrasys folk review them and update accordingly

File: platform/barefoot/sonic-platform-modules-ingrasys/debian/sonic-platform-ingrasys-s9180-32x.postinst (hunk 13 above)
Comment (mkbalani) [reply platform/barefoot/sonic-platform-modules-ingrasys/debian/sonic-platform-ingrasys-s9180-32x.postinst]: BFN build break after removing this file. Adding it back for now. We shall have ingrasys folk review them and update accordingly

File: platform/barefoot/sonic-platform-modules-ingrasys/debian/sonic-platform-ingrasys-s9180-32x.install (hunk 14 above)
Comment (mkbalani) [reply platform/barefoot/sonic-platform-modules-ingrasys/debian/sonic-platform-ingrasys-s9180-32x.install]: same comment as above

File: platform/barefoot/sonic-platform-modules-ingrasys/debian/sonic-platform-ingrasys-s9180-32x.dirs (hunk 15 above)
Comment (mkbalani) [reply platform/barefoot/sonic-platform-modules-ingrasys/debian/sonic-platform-ingrasys-s9180-32x.dirs]: same comment as above

File: platform/barefoot/sonic-platform-modules-ingrasys/debian/sonic-platform-ingrasys-s9180-32x.postrm (hunk 16 above)
Comment (mkbalani) [reply platform/barefoot/sonic-platform-modules-ingrasys/debian/sonic-platform-ingrasys-s9180-32x.postrm]: same comment as above

**Comment by lguohan** [comment]
retest this please

**Comment by lguohan** [comment]
can you fix the vs test by below patch.

```
diff --git a/platform/vs/tests/conftest.py b/platform/vs/tests/conftest.py
index 9256597..48ad7e1 100644
--- a/platform/vs/tests/conftest.py
+++ b/platform/vs/tests/conftest.py
@@ -54,8 +54,21 @@ class AsicDbValidator(object):
         atbl = swsscommon.Table(self.adb, "ASIC_STATE:SAI_OBJECT_TYPE_ACL_TABLE")
         keys = atbl.getKeys()
 
-        assert len(keys) == 1
-        self.default_acl_table = keys[0]
+        assert len(keys) >= 1
+        # Filter out DTel Acl tables
+        default_table_found = False
+        for k in keys:
+            if default_table_found:
+                break
+            (status, fvs) = atbl.get(k)
+            for item in fvs:
+                if item[0] == "SAI_ACL_TABLE_ATTR_ACL_BIND_POINT_TYPE_LIST":
+                    if 'SAI_ACL_BIND_POINT_TYPE_PORT' in item[1]:
+                        self.default_acl_table = k
+                        default_table_found = True
+                        break
+                    else:
+                        break
 
         atbl = swsscommon.Table(self.adb, "ASIC_STATE:SAI_OBJECT_TYPE_ACL_ENTRY")
         keys = atbl.getKeys()
```

**Comment by lguohan** [comment]
retest this please

**Pull request merged**